core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -133,13 +155,468 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
     return snapshotOrdinals;
   }
 
+  /**
+   * Builds a delete file index for existing deletes that were present before the start snapshot.
+   * These deletes should be applied to data files but should not generate DELETE changelog rows.
+   * Uses manifest pruning and caching to optimize performance.
+   */
+  private DeleteFileIndex buildExistingDeleteIndex(Long fromSnapshotIdExclusive) {
+    if (fromSnapshotIdExclusive == null) {
+      return DeleteFileIndex.builderFor(ImmutableList.of()).build();
+    }
+
+    Snapshot fromSnapshot = table().snapshot(fromSnapshotIdExclusive);
+    if (fromSnapshot == null) {
+      return DeleteFileIndex.builderFor(ImmutableList.of()).build();
+    }
```

Comment:
I don't believe this behavior is correct (though also not reachable).  I believe we validate that the snapshot is present at `BaseIncrementalScan.java:179`.  However, if for some reason we got to this point, you shouldn't assume that you start from the beginning of the snapshot ancestry if you cannot find the snapshot.  This should be a validation check.

Reply:
I addressed this my adding precondition check.